content/1_docs/2_cookbook/9_setup/0_understanding-oop/recipe.txt

```diff
@@ -0,0 +1,468 @@
+
+Looks a bit different than our mini example above, doesn't it? It's not only very much longer but it doesn't even start with the class definition. Let's go through what this stuff at the top means.
+
+### Namespacing
+The first line after the opening PHP tag reads:
+
+```php "/kirby/src/Cms/Page.php"
+namespace Kirby\Cms;
+```
+
+Here, we declare a (link: https://www.php.net/manual/en/language.namespaces.rationale.php text: namespace ) for the class to prevent name collisions with other classes of the same name (for example classes from third party libraries we want to use in our project) and to logically group classes that belong together. 
+
+### Importing other classes
+
+After the namespace declaration follows a long list of `use` statements.
+
+```php "/kirby/src/Cms/Page.php"
+use Kirby\Exception\Exception;
+use Kirby\Exception\InvalidArgumentException;
+use Kirby\Exception\NotFoundException;
+use Kirby\Http\Uri;
+use Kirby\Toolkit\A;
+use Kirby\Toolkit\F;
+```
+
+ The purpose of the `use` keyword here is to import classes into the file or to use aliases for class names so that we don't have to refer to them by their full names. You can read more about namespacing in the (link: https://www.php.net/manual/en/language.namespaces.importing.php text: PHP documentation).
+
+As you can see, the `Page` class relies on quite a few other classes to provide it's functionality, some of which you have probably already worked with, like the `A` class for messing with arrays or the `F` class for file handling.
+
+The next part is a comment with information about the class, followed by the actual class definition. But what is that? It doesn't just say `class Page` like in our `Book` class example above but:
+
+
+```php "/kirby/src/Cms/Page.php"
+class Page extends ModelWithContent
+```
+
+### Inheritance (`extends`)
+
+If a class extends another class, we speak of inheritance, one of the big advantages of OOP. In this case, the `Page` class is a child class of the `ModelWithContent` class and inherits all its methods and properties.
+
+However, it doesn't just inherit them but can modify existing (public or protected) methods or add additional properties and methods. Basically, the `ModelWithContent` class can be regarded as the more generic class, and `Page` is a more specific child class of this class. Other Kirby classes that inherit from `ModelWithContent` are `Site`, `File` and `User`.
+
+If we return to the `Book` example from above, it would probably make sense to extend the basic `Book` class with more specific child classes, for example for ebooks that are different from paper books. A paper book, for example, has pages you can count, an ebook doesn't. A paper book probably has a hard- or softcover, whereas an eBook has different file formats like PDF or Mobi etc. Inheritance is therefore a powerful way to modularize code while at the same time keeping code dry. You only have to redefine methods or extend where necessary.
```

Comment:
```suggestion
If we return to the `Book` example from above, it would probably make sense to extend the basic `Book` class with more specific child classes, for example for eBooks that are different from paper books. A paper book, for example, has pages you can count, an eBook doesn't. A paper book probably has a hard- or softcover, whereas an eBook has different file formats like PDF or Mobi etc. Inheritance is therefore a powerful way to modularize code while at the same time keeping code DRY ("don't repeat yourself"). You only have to redefine methods or extend where necessary.
```